File: python/data_structures/d_linked_list.py
"""
Doubly linked list version. Just for coding practice.
"""

class DoublyLinkedList:
	class Node:
		def __init__(self, value_=None, prev=None,nxt=None):
			self.value = value_
			self.prev = prev
			self.nxt = nxt

		def __str__(self):
			return f"{self.value}"

		def __repr__(self):
			# this isn't a true repr as Node can't be called outside of this class but it's good for debugging.
			return "Node({!r}, {!r}, {!r})".format(self.value, self.prev, self.nxt)

	def __init__(self,values = None):
		self._head = None
		self._length = 0
		self._tail = None
		if values:
			self._head = self.Node(values[0])
			cur = self._head
			for value in values[1:]:
				tmp = self.Node(value, cur)
				cur.nxt = tmp
				cur = cur.nxt

			self._tail = cur
			self._length = len(values)

	def __str__(self):
		cur = self._head
		output = "["
		while cur:
			output += str(cur)
			cur = cur.nxt
			if cur is not None:
				output += ", "


		output += "]"
		return output

	# def __repr__(self):
	# 	# Will actually make one later on.
	# 	//return repr(self._head)



	def __reversed__(self):
		cur = self._head
		self._head = self._tail
		self._tail =cur
		prev = None
		nxt = None
		while cur is not None:
			nxt  = cur.nxt
			cur.nxt = prev
			cur.prev = nxt
			prev = cur
			cur = nxt


	def __len__(self):
		return self._length

	def push(self, value):
		if self._head is None:
			self._head = self.Node(value)
			self._tail = self._head
		else:
			p = self.Node(value,self._tail)
			self._tail.nxt = p
			self._tail = p

		self._length += 1

	def pop(self):
		if self._head is None:
			return -1
		popped = self._tail
		if self._length == 1:
			self._head = self._tail = None
		else:
			self._tail=popped.prev
			self._tail.nxt = None

		self._length-=1
		return popped.value

	def get(self, idx):
		if idx < 0 or  idx > self._length:
			return None
		cur = self._head
		# for i in range(idx):
		# 	cur = cur.nxt
		if idx <= (self._length//2):
			i = 0
			while i != idx:
				cur = cur.nxt
				i+=1
		else:
			i = self._length - 1
			cur = self._tail
			while i != idx:
				cur = cur.prev
				i-=1
		return cur

	def insert(self, idx, value):
		if 0 > idx > self._length:
			return None
		elif self._head is None:
			self._head = self.Node(value)
		elif idx == self._length:
			self.push(value)
		elif idx == 0:
			self._head = self.Node(value, self._head)
		else:
			cur = self.get(idx)
			# need 2 pointers since Python doesn't actually
			# update the pointers the way you'd think it would.
			old_prev = cur.prev
			new_node = self.Node(value, old_prev, cur)
			old_prev.nxt = new_node
			cur.prev = new_node


		self._length += 1

	def set(self, idx, value):
		if self._length > idx < 0:
			return None
		else:
			p = self.get(idx)
			p.value = value

	def shift(self):
		if self._head:
			tmp = self._head
			self._head = tmp.nxt
			self._head.prev = None
			self._length -= 1
			if self._length == 0:
				self._tail = 0
		else:
			return None

	def unshift(self, value):
		if self._head:
			self._head = self.Node(value, None, self._head)
		else:
			self._head = self.Node(value)
			self._tail = self._head

	def remove(self, idx):
		if idx  < 0 or idx > self._length:
			return None
		if idx == self._length - 1:
			self.pop()
		elif idx == 0:
			self.shift()
		else:
			prev = self.get(idx-1)
			remove = prev.nxt
			prev.nxt = remove.nxt
			self._length-=1


	def rev(self):
		self.__reversed__()



if __name__ == "__main__":
	ll = DoublyLinkedList()
	ll.push(4)
	ll.push(5)
	ll.push(6)
	ll.insert(2,222)
	ll.push(7)
	ll.push(8)
	ll.push(9)
	ll.push(10)
	ll.rev()
	ll.pop()
	ll.set(2,100)
	ll.insert(5,555)
	ll.insert(4,444)
	ll.shift()
	ll.unshift(6)
	ll.remove(5)
	print(ll)

File: js/data_structures/d_linked_list.js
class DoublyLinkedList {
	constructor(values){
		if(arguments.length > 0) {
			this._length = values.length;
			this._head = new DoublyLinkedList.Node(values[0]);
			let current = this._head;
			for (let i = 1; i < this._length; i++) {
				current.next = new DoublyLinkedList.Node(values[i],current);
				current = current.next;
			}
			this._tail = current;
		}
		else{
			this._length = 0;
			this.head = null;
			this._tail = null;
		}
	}

	static Node = class {
		constructor(val=null,prev_=null,next_=null){
			this.value=val;
			this.prev = prev_;
			this.next=next_;
		}
	}

	push(value){
		if(!this._head){
			this._head = new DoublyLinkedList.Node(value);
			this._tail = this._head;
		}
		else{
			let p = new DoublyLinkedList.Node(value,this._tail);
			this._tail.next = p;
			this._tail = p;
		}
		this._length++;
	}

	get(idx){
		if (idx < 0 || idx> this._length){
			return undefined;
		}
		let current = this._head;
		if(idx <= (this._length >>> 1) ){
			let i = 0;
			while(i !== idx){
				current = current.next;
				i++;
			}
		}
		else{
			let i = this._length - 1;
			current = this._tail;
			while (i !== idx){
				current = current.prev;
				i--;
			}
		}

		return current;
	}

	pop(){
		if(!this._head){
			return undefined;
		}
		let tmp = this._tail;
		if(this._length === 1){
			this._head=this._tail=null;
		}
		else{
			this._tail = tmp.prev;
			this._tail.next = null;
		}
		this._length--;
		return tmp.value;
	}

	/**
	 *
	 * @param idx Number
	 * @param value mixed
	 * @returns
	 */
	insert(idx,value){
		if(idx > this._length || idx < 0)
			return undefined;
		else if(idx === 0){
			this.unshift(value);
		}
		else if(idx === this._length){
			this.push(value);
		}
		else{
			let afterNode = this.get(idx);
			let beforeNode = afterNode.prev;
			let newNode = new DoublyLinkedList.Node(value, beforeNode, afterNode);
			beforeNode.next = newNode;
			afterNode.prev = newNode;
			this._length++;
		}
	}

	set(idx,value){
		if( idx > 0 && idx < this._length){
			let p = this.get(idx);
			p.value = value;
		}
	}

	shift(){
		if(this._head){
			let tmp = this._head;
			this._head = tmp.next;
			this._length--;
			if(this._length === 0){
				this._tail = undefined;
			}
			return tmp;
		}
	}

	unshift(value){
		if(this._head){
			this._head = new DoublyLinkedList.Node(value,null,this._head);
		}
		else{
			this._head = new DoublyLinkedList.Node(value);
			this._tail = this._head;
		}
		this._length++;
	}

	remove(idx){
		if(idx < 0 || idx > this._length){
			return undefined;
		}
		else if(idx === this._length - 1){
			this.pop();
		}
		else if(idx === 0){
			this.shift();
		}
		else{
			let prev = this.get(idx-1);
			let remove = prev.next;
			prev.next = remove.next;
			this._length--;
		}
	}

	reverse(){
		let current = this._head
		this._head = this._tail
		this._tail = current

		let prev = null, next = null

		while (current !== null){
			prev = current.prev
			next = current.next
			current.prev = next
			current.next = prev
			current = next
		}
	}

	toString(){
		let current = this._head;
		let output = '['
		for(let i=0; i<this._length; i++){
			output += current.value.toString();
			if(i < this._length - 1){
				output += ', ';
			}
			current = current.next;
		}
		output += ']';
		return output;
	}
	display(){
		console.log(this.toString());
	}

}
function test(){
	const ll = new DoublyLinkedList([4,5]);
	ll.push(6);
	ll.insert(2,222);
	ll.push(7);
	ll.push(8);
	ll.push(9);
	ll.push(10);
	ll.reverse();
	ll.pop();
	ll.set(2,100);
	ll.insert(5,555);
	ll.insert(4,444);
	ll.shift();
	ll.unshift(6);
	ll.remove(5);
	const expected = '[6, 9, 100, 7, 444, 555, 222, 5]';
	if(ll.toString() === expected ){
		console.log("All's clear");
	}
	else{
		console.log(`We got ${ll.toString()} but we expected ${expected}.`)
	}
}
if (typeof module !== 'undefined' && require.main === module) {
	test();
}
